### PR TITLE
Only check server SSL certificate if requested

### LINKED ISSUE #30075: Inconsistent SSL certificate validity check using auth manager
Situation:
* Postgres server with SSL enabled and a self-signed certificate
* SSLMode=require in the QGIS connection

# Not using auth manager

QGIS successfully connects if username and password are correct

# Using auth manager basic auth

QGIS fails to connect with username and password, complaining about an "SSL error: certificate verify failed".

Screenshot from 2019-06-04 09-54-20

# Analysis

The QGIS Basic Auth Method expansion adds a "sslrootcert" part to the connection string which, in combination with the following note from [the postgres docs](https://www.postgresql.org/docs/9.1/libpq-ssl.html) trigger this behaviour.

> Note: For backwards compatibility with earlier versions of PostgreSQL, if a root CA file exists, the behavior of sslmode=require will be the same as that of verify-ca, meaning the server certificate is validated against the CA. Relying on this behavior is discouraged, and applications that need certificate validation should always use verify-ca or verify-full.

### PR DESCRIPTION
**Pull request by m-kuhn**
Only checks for a valid server ssl certificate if the auth method chosen requires this (verify-ca or verify-full).

At the moment, as soon as auth management is turned on, the `sslmode=require` is having the same effect as `sslmode=verify-ca `, making it impossible to "require" QGIS to use a self-signed certificate.

Fixes #30075

### CODE REVIEW

File: src/auth/basic/qgsauthbasicmethod.cpp [hunk 1]
```diff
@@ -105,18 +105,28 @@ bool QgsAuthBasicMethod::updateDataSourceUriItems( QStringList &connectionItems,
     return false;
   }
 
+  QString sslMode = QStringLiteral( "prefer" );
```
Comment (elpaso) [inline src/auth/basic/qgsauthbasicmethod.cpp]: Just a note : according to the PG documentation, "prefer" should be the default anyway if "sslmode" is not set.

**Comment by luipir** [comment]
LGTM

File: src/auth/basic/qgsauthbasicmethod.cpp (hunk 1 above)
Comment (m-kuhn) [reply src/auth/basic/qgsauthbasicmethod.cpp]: That's exactly why it's initialized to the default here :)

Comment (elpaso) [reply src/auth/basic/qgsauthbasicmethod.cpp]: I was just saying that you are not using that value anywhere: a few lines below you check for start with 'verify', so leaving the initial value blank would work just the same.

**Review by elpaso — APPROVED** [review]
LGTM, IIRC there were some tests for PG auth connections, you may want to add a test case there.
See:
https://github.com/qgis/QGIS/blob/master/tests/src/python/test_authmanager_ogr_postgres.py
and:
https://github.com/qgis/QGIS/blob/master/tests/src/python/test_authmanager_password_postgres.py

**Comment by m-kuhn** [comment]
I'm failing to get them to work.

Do you remember if I need to setup some local crt file (/home/mkuhn/.postgresql/root.crt)?

```
496: ======================================================================
496: FAIL: testValidAuthAccess (__main__.TestAuthManager)
496: ----------------------------------------------------------------------
496: Traceback (most recent call last):
496:   File "/home/mkuhn/dev/qgis/QGIS/tests/src/python/test_authmanager_password_postgres.py", line 225, in testValidAuthAccess
496:     self.assertTrue(pg_layer.isValid())
496: AssertionError: False is not true
496: 
496: ----------------------------------------------------------------------
496: Ran 2 tests in 3.694s
496: 
496: FAILED (failures=1)
496: 04 09:41:45.708 GMT [20091] LOG:  database system is ready to accept connections\n'
496: PostGIS(1): Connection to database failed
496: root certificate file "/home/mkuhn/.postgresql/root.crt" does not exist
496: Either provide the file or change sslmode to disable server certificate verification.
496: 
496: PostGIS(1): Connection to database failed
496: root certificate file "/home/mkuhn/.postgresql/root.crt" does not exist
496: Either provide the file or change sslmode to disable server certificate verification.
496: 
496: sys:1: ResourceWarning: unclosed file <_io.BufferedReader name=25>
496: sys:1: ResourceWarning: unclosed file <_io.BufferedReader name=27>
496: CMake Error at PyQgsAuthManagerPasswordPostgresTest.cmake:15 (MESSAGE):
496:   Test failed: 1
```

**Comment by elpaso** [comment]
@m-kuhn no, you do not need to setup any cert, that is done in the test by configuring a temporary pg server with test certs, including the root cert: https://github.com/qgis/QGIS/blob/master/tests/src/python/test_authmanager_password_postgres.py#L123

**Comment by m-kuhn** [comment]
Any other idea what could be going wrong?

**Comment by m-kuhn** [comment]
Let's see how those tests perform, they were completely written in the dark

**Review by dakcarto — APPROVED** [review]
@m-kuhn, good catch. Looks good to me (the auth code changes, that is).

File: tests/src/python/test_authmanager_password_postgres.py [hunk 2]
```diff
@@ -55,142 +51,59 @@
...
-        cls.server.terminate()
-        os.kill(cls.server.pid, signal.SIGABRT)
-        del cls.server
-        time.sleep(2)
-        rmtree(QGIS_AUTH_DB_DIR_PATH)
-        rmtree(cls.tempfolder)
+        def printMessage(tag, msg, level):
+            with open('/tmp/fmt.log', 'a') as f:
+                f.write('{}: {}'.format(tag, msg))
+        QgsApplication.instance().messageLog().messageReceived.connect(printMessage)
```
Comment (m-kuhn) [inline tests/src/python/test_authmanager_password_postgres.py]: Not to self: delete these lines

File: tests/testdata/Dockerfile-postgis [hunk 3]
```diff
@@ -0,0 +1,7 @@
+FROM kartoza/postgis-11.0-2.5
+
+ADD auth_system/certs_keys/localhost_ssl_cert.pem /etc/ssl/certs/localhost_ssl_cert.pem
+ADD auth_system/certs_keys/localhost_ssl_key.pem /etc/ssl/private/localhost_ssl_key.pem
+ADD auth_system/certs_keys/chains_subissuer-issuer-root_issuer2-root2.pem /etc/ssl/certs/chains_subissuer-issuer-root_issuer2-root2.pem
+
+ADD temp/setup-ssl.sh /
```
Comment (m-kuhn) [inline tests/testdata/Dockerfile-postgis]: Note to self: once this works, ask kartoza for a new release with https://github.com/kartoza/docker-postgis/pull/129

Comment (m-kuhn) [reply tests/testdata/Dockerfile-postgis]: CC @timlinux

File: .ci/travis/linux/docker-compose.travis.yml [hunk 4]
```diff
@@ -1,9 +1,15 @@
 version: '3'
 services:
   postgres:
-    image: kartoza/postgis:9.5-2.2
+    build:
+      dockerfile: Dockerfile-postgis
+      context: ../../../tests/testdata
     environment:
       - ALLOW_IP_RANGE="172.18.0.0/16"
+# The following files are added in Dopckerfile-postgis
```
Comment (m-kuhn) [inline .ci/travis/linux/docker-compose.travis.yml]:
```suggestion
# The following files are added in Dockerfile-postgis
```

**Comment by m-kuhn** [comment]
Looks like we are almost there with the container based SSL checks

Right now I am using the localhost keys on which the tests complain because the server runs on the hostname `postgres`

https://travis-ci.org/qgis/QGIS/jobs/541698334#L1623-L1624

@elpaso (how) can we generate a key for `postgres`? Are there other keys and we could alias postgres with another hostname? Or as a last resort we could temporarily update /etc/hosts to point localhost to postgres, but that would be a hack I would like to avoid.

**Comment by m-kuhn** [comment]
@dakcarto @elpaso do you have any idea what's going wrong in these tests?

From what I can locally here, the docker-based postgres server is setup with proper ssl key and certificate (for the hostname "postgres" which is used by the test).
However, in the tests ["SSL error: certificate verify failed"](https://travis-ci.org/qgis/QGIS/jobs/542193250#L1628) and I can see that it complains about ["root certificate file "/root/.postgresql/root.crt" does not exist"](https://travis-ci.org/qgis/QGIS/jobs/542193250#L1623). What I don't understand is, why it even looks for this, since the [`sslrootcert` parameter is provided](root certificate file "/root/.postgresql/root.crt" does not exist).

I'm happy for any input.

File: src/auth/basic/qgsauthbasicmethod.cpp [hunk 5]
```diff
@@ -272,6 +283,7 @@ bool QgsAuthBasicMethod::updateDataSourceUriItems( QStringList &connectionItems,
       else
       {
         connectionItems.append( caparam );
+        qWarning() << QStringLiteral( "Connection items after appending" ).arg( connectionItems.join( "&" ) );
```
Comment (elpaso) [inline src/auth/basic/qgsauthbasicmethod.cpp]: you're missing a %1 here

Comment (m-kuhn) [reply src/auth/basic/qgsauthbasicmethod.cpp]: It's not going to end up in the final PR and good enough to provide the required info in the travis log, no worries ;)

File: tests/src/python/test_authmanager_password_postgres.py [hunk 6]
```diff
@@ -6,18 +6,13 @@
 checks if QGIS can use a stored auth manager auth configuration to access
 a Password protected postgres.
 
-Configuration from the environment:
+It uses a docker container as postgres/postgis server with certificates from tests/testdata/auth_system/certs_keys
 
-    * QGIS_POSTGRES_SERVER_PORT (default: 55432)
-    * QGIS_POSTGRES_EXECUTABLE_PATH (default: /usr/lib/postgresql/9.4/bin)
+Use ocker-compose -f .ci/travis/linux/docker-compose.travis.yml up postgres to start the server.
```
Comment (elpaso) [inline tests/src/python/test_authmanager_password_postgres.py]:
```suggestion
Use docker-compose -f .ci/travis/linux/docker-compose.travis.yml up postgres to start the server.
```

File: tests/testdata/Dockerfile-postgis [hunk 7]
```diff
@@ -0,0 +1,9 @@
+FROM kartoza/postgis:11.0-2.5
+
+ADD auth_system/certs_keys/postgres.crt /etc/ssl/certs/postgres_cert.crt
+ADD auth_system/certs_keys/postgres.key /etc/ssl/private/postgres_key.key
+ADD auth_system/certs_keys/issuer_ca_cert.pem /etc/ssl/certs/issuer_ca_cert.pem
+
+RUN chmod 400 /etc/ssl/private/postgres_key.key
+
+ADD temp/setup-ssl.sh /
```
Comment (elpaso) [inline tests/testdata/Dockerfile-postgis]: Are you sure this setup-ssl.sh is actually run inside the container?

**Comment by elpaso** [comment]
@m-kuhn I would try to run the container locally, ssh into it and check what's going wrong, I suspect that the pg configuration does not take the following right, 

```
ssl_cert_file = '...'
ssl_key_file = '...'
ssl_ca_file = '...'
```

> @dakcarto @elpaso do you have any idea what's going wrong in these tests?
> 
> From what I can locally see here, the docker-based postgres server is setup with proper ssl key and certificate (for the hostname "postgres" which is used by the test). If I add `issuer_ca_cert.pem` to the trusted ca's in the auth system, I can connect to this postgres instance.
> If I run the tests, they connect to the server (and then complain about a missing database because it's not setup, but that's a step further already than on travis)
> 
> However, in the tests ["SSL error: certificate verify failed"](https://travis-ci.org/qgis/QGIS/jobs/542193250#L1628) and I can see that it complains about ["root certificate file "/root/.postgresql/root.crt" does not exist"](https://travis-ci.org/qgis/QGIS/jobs/542193250#L1623). What I don't understand is, why it even looks for this, since the [`sslrootcert` parameter is provided](https://travis-ci.org/qgis/QGIS/jobs/542193250#L1607).

I think that `sslrootcert` is the root cert for the client (QGIS), while the message `root certificate file "/root/.postgresql/root.crt` comes from the server.

File: tests/testdata/Dockerfile-postgis (hunk 7 above)
Comment (m-kuhn) [reply tests/testdata/Dockerfile-postgis]: Pretty sure:

* That's how the upstream package works
* It sets up ssl properly when upping this docker compose locally here
* Looking at the "/root/.postgresql/root.crt" error on travis it much more looks like a client- (test-)side problem than a server config problem

**Comment by m-kuhn** [comment]
> I think that sslrootcert is the root cert for the client (QGIS), while the message root certificate file "/root/.postgresql/root.crt comes from the server.

I don't think it comes from the server:

> To allow server certificate verification, one or more root certificates must be placed in the file ~/.postgresql/root.crt in the user's home directory. (On Microsoft Windows the file is named %APPDATA%\postgresql\root.crt.) Intermediate certificates should also be added to the file if they are needed to link the certificate chain sent by the server to the root certificates stored on the client.
> [...]
> The location of the root certificate file and the CRL can be changed by setting the connection parameters sslrootcert and sslcrl or the environment variables PGSSLROOTCERT and PGSSLCRL.

Source: https://www.postgresql.org/docs/11/libpq-ssl.html

**Comment by elpaso** [comment]
@m-kuhn yes, I think you're right. It comes from the client QGIS. I've no idea why it doesn't get to libpq. SSL is very picky about file permissions, did you check the permissions on the certs?

**Comment by m-kuhn** [comment]
Hmm... What's the requirement on the certs?

I think it's mainly (only?) picky on private keys and here there's only one PK (server) involved which gets it's mode fixed in the Dockerfile. Certs should not be confidential information.

**Comment by elpaso** [comment]
I believe 600 but I'm not sure, worths checking btw.
Another possibility is that the client part has been broken in master, if the tests were not running on travis it may have gotten unnoticed. I can try to run the tests locally.

**Comment by m-kuhn** [comment]
:tada: it was the chain cert that required to be used on the client (and which I had enabled in my local auth setup :tada:

**Comment by elpaso** [comment]
@m-kuhn beware the postgres.crt expires in one year , you might want it to last longer.

**Comment by m-kuhn** [comment]
Really? Oops, I always did it for 1000 years, maybe I failed on that in the last run... will check. thx

**Pull request merged**